crates/typst/src/layout/align.rs

```diff
@@ -435,6 +456,16 @@ impl From<OuterVAlignment> for VAlignment {
     }
 }
 
+impl From<VAlignment> for OuterVAlignment {
+    fn from(value: VAlignment) -> Self {
+        match value {
+            VAlignment::Top => OuterVAlignment::Top,
+            VAlignment::Bottom => OuterVAlignment::Bottom,
+            VAlignment::Horizon => unreachable!(),
+        }
```

Comment:
The last path isn't unreachable; we should use a fallible trait instead (returning a Result), as user code can trigger this path (in theory).

In fact, we only ever rely on FromValue anyway, which is perhaps a bit odd. One possible approach here is to switch to `TryFrom<HAlignment/VAlignment>` everywhere, and then have the FromValue implementations just call the TryFrom impl and propagate the error up.

Reply:
I chose `unreachable!()` here because I think user code cannot trigger this path, as the `Horizon` value is already striken down by the `FromValue` trait's impl for `SpecificAlignment<H, V>`, preventing the error from propagating this far. WDYT?

Reply:
Let me elaborate my view a bit further.

You generally only use `unreachable!()` when there are internal invariants which you are sure are always upheld when the code that may lead to the unreachable!() is used. The code above does not appear to be used anywhere, so no logic is being upheld or let alone checked here. Not just that, but even if we were using it at FromValue as you said, that still doesn't stop us from using this From implementation elsewhere in the codebase without upholding the invariant, by accident. In other words, this From impl is fully detached - both visually and logically - from the logic which checks whether or not it would be valid to use its code, making it overall a worse experience to use it.

It's true that you could use Rust's `unsafe fn` syntax to indicate that the function requires an invariant to be upheld to be safely called (note that `unsafe fn` is conceptually distinct from a function using unsafe code). It's also true that you could just add a doc comment ensuring code calling this From shouldn't ever provide an invalid value, lest you get a panic. But all of those options (including the current one) incur extra mental overhead for the devs where it isn't needed - what determines whether or not this implementation should be called (that is, whether or not its invariant is upheld) **always** comes from user code. In other words, every single place where we would use From we would have to perform the same check (if v isn't Horizon => ...). So we shouldn't treat a Horizon value here as an internal logic bug and panic, forcing callers to make the check - rather, we should just switch to a fallible trait which returns an Err when the user specifies an invalid value. With that approach, **we can remove all such checks everywhere they occur** (otherwise they'd be necessarily duplicated) and just rely on the single check implementation provided by TryFrom. By removing the From implementation, we completely eliminate the possibility of misusing it - we ensure the check always occurs -, while at the same time reduce duplication when using it - no need to repeat the check everywhere. 

The main point here is that, to safely use this From, you will have to check anyway, since otherwise it would turn a user error into an internal logic error. So ideally we'd just make it so the only way possible to use this From _is_ by running the check, because that's the only way it's implemented. We do that by using TryFrom (that's its purpose: a trait for conversions which might fail, depending on external logic).

Hope that clears it up.